auth0/v3/authentication/token_verifier.py

```diff
@@ -0,0 +1,225 @@
+        self._fetcher = _jwks_fetcher or JwksFetcher(jwks_url)
+
+    def _fetch_key(self, key_id=None):
+        return self._fetcher.get_key(key_id)
+
+
+class JwksFetcher():
+    CACHE_TTL = 600  # 10 min cache lifetime
+
+    def __init__(self, jwks_url, cache_ttl=CACHE_TTL):
+        self._jwks_url = jwks_url
+        self._init_cache(cache_ttl)
+        return
+
+    def _init_cache(self, cache_ttl):
+        self._cache_value = {}
+        self._cache_date = 0
+        self._cache_ttl = cache_ttl
+        self._cache_is_fresh = False
+
+    def _fetch_jwks(self, force=False):
+        has_expired = self._cache_date + self._cache_ttl < time.time()
+
+        if not force and not has_expired:
+            # Return from cache
+            self._cache_is_fresh = False
+            return self._cache_value
+
+        # Invalidate cache and fetch fresh data
+        self._cache_value = {}
+        response = requests.get(self._jwks_url)
+
+        if response.ok:
+            # Update cache
+            jwks = response.json()
+            self._cache_value = self._parse_jwks(jwks)
+            self._cache_is_fresh = True
+            self._cache_date = time.time()
+        return self._cache_value
+
+    @staticmethod
+    def _parse_jwks(jwks):
+        keys = {}
+
+        for key in jwks['keys']:
+            # noinspection PyUnresolvedReferences
+            # requirement already includes cryptography -> pyjwt[crypto]
+            rsa_key = jwt.algorithms.RSAAlgorithm.from_jwk(json.dumps(key))
+            keys[key["kid"]] = rsa_key
+        return keys
+
+    def get_key(self, key_id):
+        keys = self._fetch_jwks()
+
+        if keys and key_id in keys:
+            return keys[key_id]
+
+        if not self._cache_is_fresh:
+            keys = self._fetch_jwks(force=True)
+            if keys and key_id in keys:
+                return keys[key_id]
+        raise TokenValidationError('RSA Public Key with ID "{}" was not found.'.format(key_id))
+
+
+class TokenVerifier():
+
+    def __init__(self, signature_verifier, issuer, audience, leeway=0, _clock=None):
+        if not signature_verifier or not isinstance(signature_verifier, SignatureVerifier):
+            raise TypeError("signature_verifier must be an instance of SignatureVerifier.")
+
+        self._options = {
+            'signature_verifier': signature_verifier,
+            'iss': issuer,
+            'aud': audience,
+            'leeway': leeway,
+            '_clock': _clock
+        }
+
+    def verify(self, token, nonce=None, max_age=None):
+        # Verify token presence
+        if not token or type(token) != str:
+            raise TokenValidationError("ID token is required but missing.")
+        if nonce and type(nonce) != str:
+            raise TokenValidationError("Nonce not specified.")
+        if max_age and type(max_age) != int:
+            raise TokenValidationError("Max Age not specified.")
+
+        opt = self._options.copy()
```

Comment:
I did this only to avoid having to do `self._options.property` every time I wanted to access `property`. Now I can do `opt.property`

Reply:
You don't really need an options dictionary, you can use class attributes and the values as they were passed in to verify:

```python
class Verifier():
  __init__(self, foo):
    self.foo = foo

  verify(self, payload, bar):
    if payload['foo'] !== self.foo:
      raise ...
    if payload['bar'] !== bar:
      raise ...
```

Reply:
Thanks adam. 
Researching a bit, I found the `dict.copy()` function helpful here. I also could use the `dictionary.get(key, defaultValue)` method which accepts a default value in case the key is missing. I'll try to update the code using something of that to avoid having many class attributes. I don't know if there's a limit on class attributes (will find out), but the main reason I used an options object was to avoid prefixing everything with `self.`.

Reply:
I tried the class attributes and they look much cleaner. Thanks for insisting 👍